### ISSUE TITLE
Bug:  issue in useEffect in React Code ? 

### ISSUE THREAD

**Issue by sk16** [facebook/react#26923]
I’m having an issue with buggy code in my progress bar. The transition doesn’t start unless I wrap setStartTransition within setTimeout or requestAnimationFrame, as shown in this example: 
https://codesandbox.io/s/progress-bar-buggy-3pby7c?file=/src/progressBar.js. 

However, my understanding is that calling setStartTransition(true) after render should start the transition without the need for a timeout . 
fixed code : https://codesandbox.io/s/progress-bar-99nxwl?file=/src/progressBar.js. 

I think there is an issue with useEffect.Triggering  setStartTransition inside useEffect , is behaving as if triggering in quick succession. and onTransition event handler does'nt get called

## Steps To Reproduce

Kindly refer to this codesandbox code : https://codesandbox.io/s/progress-bar-buggy-3pby7c?file=/src/progressBar.js. 

Link to code example:

https://codesandbox.io/s/progress-bar-buggy-3pby7c?file=/src/progressBar.js

## The current behavior

As per above codesandbox link, transition should get started (All of progress bar should be filled ) . In chrome this issue happens  
intermittently . But in  other browser , this issue is consistent.

## The expected behavior

Triggering  setStartTransition in useEffect , transition should get started, but to start I have to wrap with setTimeout : https://codesandbox.io/s/progress-bar-99nxwl?file=/src/progressBar.js

**Comment by sk16** [comment]
here  , what is happening , progressBar renders in DOM (base of progress bar is rendered) - > setTransition(true) get invoked , progressBar rerenders with transition in effect . Ideally code should have run. but its not.  @AsheerAF911  can you explain it step by step

**Comment by AsheerAF911** [comment]
The problem here is of not re-rendering. Your progressBar re-renders but CSS transition is not applied properly. Below is the reason:
1. Immediate setting of 'startTransition' to true: In the original code, when 'startFilling' is true, the 'startTransition' state variable is immediately set to true without any delay. This means that the transition class is applied immediately, without allowing the component to render first. As a result, the CSS transition doesn't work as intended.

2. Missing delay for transition: Since 'startTransition' is set to true immediately in the original code, there is no delay specified for the CSS transition. The 'transitionDuration' property in the 'reactStyle' object is set correctly based on the duration prop, but the actual transition doesn't happen due to the immediate application of the transition class.  

In conclusion you need wrap 'setStartTransition' within 'setTimeout' or 'requestAnimationFrame' so that there is delay between re-render of our function and application of CSS transition.

I hope your doubt is cleared, if there is still any doubt left feel free to ask.
